public/app/features/alerting/unified/components/rule-editor/query-and-alert-condition/reducer.ts

```diff
@@ -18,6 +19,34 @@ export interface QueriesAndExpressionsState {
   queries: AlertQuery[];
 }
 
+const findDataSourceFromExpressionRecursive = (
```

Comment:
👏

Reply:
Can we move this to a util file? I feel like it doesn't belong in the reducer and could be useful in other parts of alerting.

Reply:
Good idea, I will move it!